src/main/scala/chisel3/util/Inline.scala

```diff
@@ -0,0 +1,152 @@
+  *   Seq(x, y, z).map(_.io.a := io.a)
+  * }
+  * }}}
+  */
+trait FlattenInstance { self: BaseModule =>
+  chisel3.experimental.annotate(
+    new ChiselAnnotation with RunFirrtlTransform {
+      def toFirrtl = FlattenAnnotation(self.toNamed)
+      def transformClass = classOf[Flatten] } )
+  chisel3.experimental.annotate(
+    new ChiselAnnotation {
+      def toFirrtl = NoDedupAnnotation(self.toNamed) } )
+}
+
+/** Provide methods for controlling inlining and flattening from within a
+  * [[chisel3.experimental.BaseModule]].
+  *
+  * Inlining means that an instance of a module will be replaced with the
+  * contents of the module. Instances inside an inlined instace will be
+  * moved into the module doing the inlining, but will not also be
+  * inlined.
+  *
+  * Flattening means that an instance of a module will be recursively
+  * inlined. All instances inside a flattened instance will also be
+  * inlined.
+  */
+trait InlineHelpers { self: BaseModule =>
+  /** Inline a specific instance inside a Module
+    *
+    * @param m the module to inline
+    * @example {{{
+    * /* Providing some module internals */
+    * trait Internals { this: Module =>
+    *   val io = IO(new Bundle{ val a = Input(Bool()) })
+    * }
+    * class Bar extends Module with Internals with InlineHelpers
+    * object Bar {
+    *   def apply(): Bar = Module(new Bar)
+    * }
+    * class Top extends Module with InlineHelpers {
+    *   val io = IO(new Bundle{ val a = Input(Bool()) })
+    *   val Seq(x, y, z) = Seq.fill(3)(Bar())
+    *   Seq(x, y, z).map(_.io.a := io.a)
+    *   /* Only the insance x will be inlined */
+    *   inlineComponent(x)
+    * }
+    * }}}
+    */
+  def inlineComponent(m: BaseModule): Unit = chisel3.experimental.annotate(
+    new ChiselAnnotation with RunFirrtlTransform {
+      def toFirrtl = InlineAnnotation(ComponentName(m.instanceName, self.toNamed))
+      def transformClass = classOf[InlineInstances] }
+  )
+
+  /** Flatten a specific instance inside a Module
+    *
+    * @param m the module to inline
+    * @example {{{
+    * trait Internals { this: Module =>
+    *   val io = IO(new Bundle{ val a = Input(Bool()) })
+    * }
+    * class Foo extends Module with Internals {
+    *   val bar = Module(new Bar)
+    *   bar.io.a := io.a
+    * }
+    * class Bar extends Module with Internals with InlineHelpers {
+    *   val baz = Module(new Baz)
+    *   baz.io.a := io.a
+    *   flattenComponent(baz) // All Baz inside Bar will be flattened
+    * }
+    * class Baz extends Module with Internals
+    *   // The resulting instances will be:
+    *   //   - Top
+    *   //   - Top.y
+    *   //   - Top.y.bar
+    * class Top extends Module with Internals with InlineHelpers {
+    *   val x = Module(new Foo)
+    *   val y = Module(new Foo)
+    *   Seq(x, y).map(_.io.a := io.a)
+    *   flattenComponent(x)
+    * }
+    * }}}
+    */
+  def flattenComponent(m: BaseModule): Unit = chisel3.experimental.annotate(
```

Comment:
I don't love this API. It's not the worst thing in the world (it's similar to dontTouch) so I don't think it's worth blocking over but similarly to prefering some sort of `inlineInstances { }`-like API maybe there could be a `flattenSubModules { }` or something, of course it would require support in Chisel core

Reply:
If we keep it, I think we should make it return the argument itself and probably rename it to `flattenHierarchy` or something which IMO is more clear about what it's doing

Reply:
I'll remove this as well. If a user wants this behavior, they can write their own annotator. Otherwise, we should be absolutely certain that the API we expose publicly is the one we want to support. Keeping this back in favor of a future _better_ API makes more sense than pushing this out and then deprecating or supporting both.